tests/core/data/test_tensordense.py

```diff
@@ -0,0 +1,234 @@
+        pytest.param(_valid_list(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_list(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_list(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_list(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_list(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_list(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_list(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_list(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+        pytest.param(_valid_tensor(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_tensor(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_tensor(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+    ])
+    def test_init_from_wrong_input(self, arg, kwargs, error):
+        """
+        Test that the __init__ method raises a suitable error when passed
+        incorrectly formatted inputs.
+
+        This test also serves as a *partial* check that Dense safely handles
+        deallocation in the presence of exceptions in its __init__ method.  If
+        the tests segfault, it's quite likely that the memory management isn't
+        being done correctly in the hand-off us setting our data buffers up and
+        marking the numpy actually owns the data.
+        """
+        with pytest.raises(error):
+            DenseTensor(arg, **kwargs)
+
+    def test_copy_returns_a_correct_copy(self, data_tensor_dense):
+        """
+        Test that the copy() method produces an actual copy, and that the
+        result represents the same matrix.
+        """
+        original = data_tensor_dense
+        copy = data_tensor_dense.copy()
+        assert original is not copy
+        assert np.all(original._tf == copy._tf)
```

Comment:
You should also check that mutations of the original don't affect the copy, and vice-versa.

Reply:
What do you mean by mutations? Could you provide me an example with when this could go wrong?  

If by mutations you mean that I should check that an "in place" modification of original._tf does no change copy._tf (for example by doing something like: `original._tf[0, 0] += 1`) then, it turns out that for `Tensor` it is not possible to modify it in place. For instance, `original._tf[0, 0] += 1` will return `TypeError`. `Variables` from TensorFlow can but I plan to implement them in another class.

In any case, thinking about your comment I realised that this can happen:
```python
original = data_tensor_dense
maybe_copy = DenseTensor(original._tf, copy=False) # This is not what happens with copy() method but could happen!
print(original is not maybe_copy) # True
print(np.all(original._tf == maybe_copy._tf)) #True
print(maybe_copy._tf is not original._tf) # False!!!!!!
```
So I will add in the tests: `assert copy._tf is not original._tf`. In any case I am not sure if this would matter that much as tensors can not be modified "in place".